eng/DocsGenerator/README.template.md

```diff
@@ -0,0 +1,186 @@
+    workingDirectory:
+```
+
+Notice how some of the properties are only valid in a specific combination with other.
+With Sharpliner, we remove some of this complexity using nice fluent APIs:
+
+[!code-csharp[](tests/Sharpliner.Tests/AzureDevOps/DocsTests.cs#L79-L91)]
+
+### Useful macros
+Some very common pipeline patterns such as comparing the current branch name or detecting pull requests are very cumbersome to do in YAML (long conditions full of complicated `${{ if }}` syntax).
+For many of these, we have handy macros so that you get more readable and shorter code.
+
+For example this YAML
+
+[!code-yaml[](tests/Sharpliner.Tests/AzureDevOps/DocsTests.cs#L112-L118)]
+
+can become this C#
+
+[!code-csharp[](tests/Sharpliner.Tests/AzureDevOps/DocsTests.cs#L103-L106)]
+
+### Re-usable pipeline blocks
+Sharpliner lets you re-use code more easily than YAML templates do.
+Apart from obvious C# code re-use, you can also define sets of C# building blocks and re-use them in your pipelines:
+
+[!code-csharp[](tests/Sharpliner.Tests/AzureDevOps/DocsTests.cs#L122-L133)]
+
+You can then reference this library in between build steps and it will get expanded into the pipeline's YAML:
+
+[!code-csharp[](tests/Sharpliner.Tests/AzureDevOps/DocsTests.cs#L143-L153)]
+
+More about this feature can be found [here (DefinitionLibraries.md)](https://github.com/sharpliner/sharpliner/blob/main/docs/AzureDevOps/DefinitionLibraries.md).
+
+
+### Sourcing scripts from files
+When you need to add cmd, PowerShell or bash steps into your pipeline, maintaining these bits inside YAML can be error prone.
+With Sharpliner you can keep scripts in their own files (`.ps1`, `.sh`..) where you get the natural environment you're used to such as syntax highlighting.
+Sharpliner gives you APIs to load these on build time and include them inline:
+
+[!code-csharp[](tests/Sharpliner.Tests/AzureDevOps/DocsTests.cs#L168-L175)]
+
+### Correct variable/parameter types
+Frequent struggle people have with Azure pipelines is using the [right type of variable](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#understand-variable-syntax) in the right context.
+Be it a `${{ compile time parameter }}`, a `variable['used in runtime']` or a `$(macro)` syntax, with Sharpliner you won't have to worry about which one to pick as it understands the context and selects the right one for you.
+
+### Pipeline validation
+Your pipeline definition can be validated during publishing and you can uncover issues, such as typos inside `dependsOn`, you would only find by trying to run the pipeline in CI.
+This gives you a faster dev loop and greater productivity.
+
+We are continuously adding new validations as we find new error-prone spots.
+Each validation can be individually configured/silenced in case you don't wish to take advantage of these:
+
+[!code-csharp[](tests/Sharpliner.Tests/AzureDevOps/DocsTests.cs#L181-L198)]
```

Comment:
Looks really good but maybe we should just give it a named tag and look for that? I am afraid that it will be quite painful to keep moving these all the time in cases like we will need to add a `using` into the target file and now all blocks are one line lower.

So maybe we do  
`[!code-csharp[](tests/Sharpliner.Tests/AzureDevOps/DocsTests.cs#ValidationConfiguration)]`  
and then we will have

```csharp
// #ValidationConfiguration
...
// #ValidationConfiguration
```

Reply:
I wonder if it would be clearer to follow docfx's pattern of having the named tag point to a `#region <name>`

Reply:
These could be regions too, yeah